mono/metadata/class-accessors.c

```diff
@@ -424,6 +424,34 @@ mono_class_get_weak_bitmap (MonoClass *klass, int *nbits)
 	return prop->bits;
 }
 
+/**
+ * mono_class_set_failure:
+ * \param klass class in which the failure was detected
+ * \param ex_type the kind of exception/error to be thrown (later)
+ * \param ex_data exception data (specific to each type of exception/error)
+ *
+ * Keep a detected failure informations in the class for later processing.
+ * Note that only the first failure is kept.
+ *
+ * LOCKING: Acquires the loader lock.
+ */
+gboolean
+mono_class_set_failure (MonoClass *klass, MonoErrorBoxed *boxed_error)
+{
+	g_assert (boxed_error != NULL);
+
+	if (mono_class_has_failure (klass))
```

Comment:
This is racy, but I am not sure it matters. We check outside the lock for flag and then set inside without double checking.

Reply:
Agreed.

I think the harm from the race is just that we will sometimes return `TRUE` (ie "this call to set failiure was the one that won") when we should return `FALSE`.   The boxed error is allocated in a mempool so it's not leaking.  and `has_failure` always goes from 0 to 1 so it doesn't matter which thread wins.

As for the return value... I'm not seeing us check it... ever.